x/ecocredit/server/core/msg_add_credit_type.go

```diff
@@ -0,0 +1,38 @@
+package core
+
+import (
+	"context"
+
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	govtypes "github.com/cosmos/cosmos-sdk/x/gov/types"
+
+	ecocreditv1 "github.com/regen-network/regen-ledger/api/regen/ecocredit/v1"
+	"github.com/regen-network/regen-ledger/x/ecocredit/core"
+)
+
+// AddCreditType adds a new credit type to the network.
+func (k Keeper) AddCreditType(ctx context.Context, req *core.MsgAddCreditType) (*core.MsgAddCreditTypeResponse, error) {
+	if k.authority.String() != req.Authority {
+		return nil, govtypes.ErrInvalidSigner.Wrapf("invalid authority: expected %s, got %s", k.authority, req.Authority)
+	}
+
+	found, err := k.stateStore.CreditTypeTable().HasByName(ctx, req.CreditType.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	if found {
+		return nil, sdkerrors.ErrConflict.Wrapf("credit type with %s name already exists", req.CreditType.Name)
+	}
```

Comment:
we should either do this for both name and abbreviation, or just attempt the `insert` call and check the returned error for which type of failure occurred so we can supply a more descriptive error message.

for duplicate name, you'd get an unique key violation error, and for duplicate abbreviation you'll get PrimaryKeyConstraint error. so we can either check for these or just do the pattern you've done above with the addition of abbreviation check